src/ert/com_protocol/_state_machine.py

```diff
@@ -0,0 +1,233 @@
+
+        if ensemble_id not in self.state.ensembles:
+            return 0
+        return sum(
+            self.state.ensembles[ensemble_id].realizations[id_real].status
+            == pb2.STEP_SUCCESS
+            for id_real in self.state.ensembles[ensemble_id].realizations
+        )
+
+    class IllegalStateUpdate(Exception):
+        def __init__(self, reason: str):
+            super().__init__(reason)
+            self.reason: str = reason
+
+    @singledispatchmethod
+    def _update(
+        self,
+        _: Union[pb2.Job, pb2.Step, pb2.Realization, pb2.Ensemble, pb2.Experiment],
+    ) -> None:
+        pass
+
+    @_update.register
+    def _(self, job: pb2.Job) -> None:
+        logger.debug(
+            "Updating job id %d status %s from pbuf!",
+            job.id.index,
+            pb2.JobStatus.Name(job.status),
+        )
+        old_job = self._state_handle.get_job(job.id)
+        old_job.MergeFrom(job)
+
+    @_update.register
+    def _(self, real: pb2.Realization) -> None:
+        logger.debug(f"Updating realization id {real.id.realization} from pbuf!")
+        old_real = self._state_handle.get_realization(real.id)
+        old_real.MergeFrom(real)
+
+    @_update.register
+    def _(self, step: pb2.Step) -> None:
+        logger.debug(
+            f"Updating step id {step.id.step} status "
+            f"{pb2.StepStatus.Name(step.status)} from pbuf!",
+        )
+        old_real = self._state_handle.get_step(step.id)
+        old_real.MergeFrom(step)
+        # if step==success then set realization=success
+        if step.status == pb2.STEP_SUCCESS:
+            self._update(
+                pb2.Realization(id=step.id.realization, status=pb2.STEP_SUCCESS),
+            )
+
+    @_update.register
+    def _(self, ens: pb2.Ensemble) -> None:
+        logger.debug(
+            f"Updating step id {ens.id.id} status"
+            f"{pb2.EnsembleStatus.Name(ens.status)} from pbuf!",
+        )
+        old_ens = self._state_handle.get_ensemble(ens.id)
+        old_ens.MergeFrom(ens)
+
+    @_update.register
+    def _(self, exp: pb2.Experiment) -> None:
+        logger.debug(f"Updating experiment id {exp.id.id} from pbuf!")
+        old_exp = self._state_handle.get_experiment(exp.id)
+        old_exp.MergeFrom(exp)
+
+    async def update(
+        self,
+        msg: Union[
+            pb2.Job,
+            pb2.Step,
+            pb2.Realization,
+            pb2.Ensemble,
+            pb2.Experiment,
+            pb2.DispatcherMessage,
+        ],
+    ) -> None:
+        try:
+            if isinstance(msg, pb2.DispatcherMessage):
+                self._update(getattr(msg, str(msg.WhichOneof("object"))))
+            else:
+                self._update(msg)
+        except Exception:
+            logger.error(f"Failed state machine update! Current state: {self.state}")
+            raise
+
+    @property
+    def state(self) -> pb2.Experiment:
+        assert self._state_handle._experiment is not None
```

Comment:
Should we not give an "unitialized state" or something when this is None, instead of an AssertionError?